src/colorizer/utils/texture_utils.ts

```diff
@@ -1,29 +1,54 @@
 import { DataTexture } from "three";
 
-import { FeatureDataType, featureTypeSpecs } from "../types";
+import { FeatureArrayType, FeatureDataType, featureTypeSpecs } from "../types";
 
 /**
- * Calculate the squarest possible texture that `data` can fit into, extend it with
- * `emptyVal` to fit perfectly into that space, and return a tuple of `[width, height]`
+ * Calculate the squarest possible texture that `data` can fit into and return a tuple of `[width, height]`
  */
-function fitIntoSquare<T>(data: T[], emptyVal: T): [number, number] {
+function getSquarestTextureDimensions<T extends FeatureDataType>(
+  data: FeatureArrayType[T] | number[]
+): [number, number] {
   const width = Math.ceil(Math.sqrt(data.length));
   const height = Math.ceil(data.length / width);
-  const length = width * height;
-
-  while (data.length < length) {
-    data.push(emptyVal);
-  }
 
   return [width, height];
 }
 
+/**
+ * Returns a copy of the typed array `data` padded with `emptyVal` to the specified length.
+ */
+function padToLength<T extends FeatureDataType>(
+  data: FeatureArrayType[T],
+  type: T,
+  emptyVal: number,
+  length: number
+): FeatureArrayType[T] {
+  const buffer = new featureTypeSpecs[type].ArrayConstructor(length);
+  buffer.fill(emptyVal, data.length, length);
+  buffer.set(data, 0);
+  return buffer;
+}
+
 /** Pack a 1d array of data into the squarest 2d texture possible */
-export function packDataTexture(data: number[], type: FeatureDataType): DataTexture {
-  const [width, height] = fitIntoSquare(data, 0);
+export function packDataTexture<T extends FeatureDataType>(data: FeatureArrayType[T] | number[], type: T): DataTexture {
+  const [width, height] = getSquarestTextureDimensions(data);
+  const length = width * height;
+
+  // If provided a regular array of numbers, convert to TypedArray
+  if (Array.isArray(data)) {
+    data = new featureTypeSpecs[type].ArrayConstructor(data);
+  }
 
   const spec = featureTypeSpecs[type];
-  const buffer = new spec.ArrayConstructor(data);
+  const buffer = padToLength(data, type, 0, length);
```

Comment:
If `data` is a regular JS array, it gets copied into a typed array twice here, since `padToLength` has to create a new buffer with the desired length and copy in the data. It looks like the method it uses to do this (`TypedArray.set`) can accept a JS array as well. Could we forego the first copy and extend the type signature of `padToLength` to accept a regular array too? (If so, `padToLength` might need a new name to reflect the extra responsibility, like `intoPaddedTypedArray` or something.)

Reply:
This turned out to be easier than expected!